### PR TITLE
Flex Attention + Packing with BlockMask support

### PR DESCRIPTION
**Pull request by bursteratom**
# Description

~Awaiting upstream: https://github.com/huggingface/transformers/pull/36103~

Merged upstream in new PR: https://github.com/huggingface/transformers/pull/36643

### CODE REVIEW

**Comment by bursteratom** [comment]
https://github.com/huggingface/transformers/pull/36643 got merged!

File: src/axolotl/utils/config/models/input/v0_4_1/__init__.py [hunk 1]
```diff
@@ -825,6 +825,7 @@ class AxolotlInputConfig(
     xformers_attention: Optional[bool] = None
     sdp_attention: Optional[bool] = None
     s2_attention: Optional[bool] = None
+    flex_attention: Optional[bool] = None
     flash_attention: Optional[bool] = None
```
Comment (winglian) [inline src/axolotl/utils/config/models/input/v0_4_1/__init__.py]: add a pydantic validation that only one of these attention types are set

**Comment by winglian** [comment]
Discussed with Sunny, hold on this until transformers cuts a new release with the upstreamed fix. In the meantime, this PR should add some smoke tests for e2e and point to transformers@main to verify the fixes.

File: src/axolotl/utils/config/models/input/v0_4_1/__init__.py (hunk 1 above)
Comment (bursteratom) [reply src/axolotl/utils/config/models/input/v0_4_1/__init__.py]: Done

**Comment by winglian** [comment]
This still needs a fix to include multi-GPU support. 
Screenshot 2025-03-19 at 7 45 52 PM

**Comment by winglian** [comment]
Sunny reported that with multigpu, it errors because the causal mask ends up on a different device than the document and padding masks. However, it happens for llama 3.1 8B, but not for llama 3.2 1B.

image-1

https://github.com/huggingface/transformers/blob/3249c5dc1560dace3c31cdbe4797b6c878ab47de/src/transformers/integrations/flex_attention.py#L74-L127

/cc @joecummings @drisspg

**Comment by drisspg** [comment]
Compiled DDP was fixed relatively recently: https://github.com/pytorch/pytorch/issues/142313
Can you double check w/ a nightly version of PT?

File: src/axolotl/core/trainer_builder.py [hunk 2]
```diff
@@ -891,7 +891,11 @@ def build_collator(
             if "max_length" in kwargs:
                 kwargs.pop("max_length")
         elif use_batch_sampler_collator:
-            if self.cfg.model_config_type in SUPPORTED_MULTIPACK_MODEL_TYPES or (
+            if self.cfg.flex_attention is True:
```
Comment (salmanmohammadi) [inline src/axolotl/core/trainer_builder.py]:
```suggestion
            if self.cfg.flex_attention:
```

File: src/axolotl/monkeypatch/attention/flex_attn.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+                    # Create a new instance if one doesn't already exist
+                    cls._instance = super().__new__(cls)
+                return cls._instance
+
+            @torch.compiler.disable(recursive=False)
+            def __init__(self):
+                """
+                Initialize or update the singleton instance.
+                """
+                if self._is_flex_compiled is False:
```
Comment (salmanmohammadi) [inline src/axolotl/monkeypatch/attention/flex_attn.py]:
```suggestion
                if not self._is_flex_compiled:
```

File: src/axolotl/utils/schemas/config.py [hunk 4]
```diff
@@ -1250,6 +1267,26 @@ def check_adopt_torch_version(cls, data):
                 )
         return data
 
+    @model_validator(mode="before")
+    @classmethod
+    def check_flex_torch_version(cls, data):
+        if (data.get("flex_attention") is not None) and (
+            data.get("flex_attention") is True
```
Comment (salmanmohammadi) [inline src/axolotl/utils/schemas/config.py]:
```suggestion
            data.get("flex_attention")
```

**Comment by salmanmohammadi** [comment]
LGTM, did you compare training speed with and without flex attention+sample packing?

**Comment by bursteratom** [comment]
@SalmanMohammadi Yeah I did a while back and find that flex attention has similar speed with flash attention!

File: src/axolotl/monkeypatch/attention/flex_attn.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    if int(version_parts[0]) == 4:
+        if int(version_parts[1]) < 50:
+            return True
+        if int(version_parts[1]) == 50 and int(version_parts[2]) <= 99:
+            return True
+    return False
+
+
+def patch_flex():
+    is_torch_2_6 = torch.__version__.startswith("2.6")
```
Comment (salmanmohammadi) [inline src/axolotl/monkeypatch/attention/flex_attn.py:8]: nit: we have `from axolotl.utils import get_pytorch_version` for checking the torch version

File: src/axolotl/monkeypatch/attention/flex_attn.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+    # Check if the version is less than or equal to 4.50
+    if int(version_parts[0]) < 4:
+        return True
+    if int(version_parts[0]) == 4:
+        if int(version_parts[1]) < 50:
+            return True
+        if int(version_parts[1]) == 50 and int(version_parts[2]) <= 99:
+            return True
+    return False
```
Comment (salmanmohammadi) [inline src/axolotl/monkeypatch/attention/flex_attn.py]: nit
```suggestion
    return transformers.__version__ < "4.51"
```

File: src/axolotl/monkeypatch/attention/flex_attn.py (hunk 5 above)
Comment (salmanmohammadi) [reply src/axolotl/monkeypatch/attention/flex_attn.py:8]: Could also just do `is_torch_2_6 = torch.__version__ == 2.6.0`

File: src/axolotl/utils/models.py [hunk 7]
```diff
@@ -877,7 +877,19 @@ def set_attention_config(self) -> None:
...
-        if self.cfg.flash_attention:
+        if self.cfg.flex_attention:
+            self.model_kwargs["attn_implementation"] = "flex_attention"
+            self.model_config._attn_implementation = (  # pylint: disable=protected-access
+                "flex_attention"
+            )
+            from axolotl.monkeypatch.attention.flex_attn import patch_flex
+
+            patch_flex()
+            # Multigpu training with flex on torch 2.5 will need torch_compile=True so it doesn't throw error
```
Comment (salmanmohammadi) [inline src/axolotl/utils/models.py]: Why is this?

File: src/axolotl/monkeypatch/attention/flex_attn.py (hunk 5 above)
Comment (bursteratom) [reply src/axolotl/monkeypatch/attention/flex_attn.py:8]: `is_torch_version` would be a bit more involved here as I am pinning against 2.6.x , if we use `is_torch_version` this would amount to `is_torch_version("<", "2.7.0") and is_torch_version(">=", "2.6.0")`. Seems cleaner to just do `torch.__version__.startswith("2.6")`, unless there is some security reason that we don't want to invoke `torch.__version__`

File: src/axolotl/monkeypatch/attention/flex_attn.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+                return cls._instance
+
+            @torch.compiler.disable(recursive=False)
+            def __init__(self):
+                """
+                Initialize or update the singleton instance.
+                """
+                if not self._is_flex_compiled:
+                    self._compiled_flex_attention = torch.compile(
+                        flex_attention, dynamic=False, mode="max-autotune-no-cudagraphs"
```
Comment (drisspg) [inline src/axolotl/monkeypatch/attention/flex_attn.py]: setting full_graph=True is a good here as well

File: src/axolotl/monkeypatch/attention/flex_attn.py (hunk 5 above)
Comment (salmanmohammadi) [reply src/axolotl/monkeypatch/attention/flex_attn.py:8]: There won't be a 2.6.X release, the next torch release is 2.7, so we can just check for == 2.6.0

File: src/axolotl/monkeypatch/attention/flex_attn.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
+"""Flex attention monkey patch"""
+
+import torch
+import transformers
+
+
+def patch_flex():
+    is_torch_2_6 = torch.__version__ == "2.6.0"
+    is_transformers_below_4_51 = transformers.__version__ <= "4.51.0"
```
Comment (salmanmohammadi) [inline src/axolotl/monkeypatch/attention/flex_attn.py]:
```suggestion
    is_transformers_below_4_51 = transformers.__version__ < "4.51.0"
```
right?

File: src/axolotl/monkeypatch/attention/flex_attn.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+                return cls._instance
+
+            @torch.compiler.disable(recursive=False)
+            def __init__(self):
+                """
+                Initialize or update the singleton instance.
+                """
+                if not self._is_flex_compiled:
+                    self._compiled_flex_attention = torch.compile(
+                        flex_attention, dynamic=False, mode="max-autotune-no-cudagraphs"
```
Comment (winglian) [inline src/axolotl/monkeypatch/attention/flex_attn.py]:
```suggestion
                        flex_attention,
                        dynamic=False,
                        mode="max-autotune-no-cudagraphs",
                        fullgraph=True,
```

**Review by winglian — APPROVED** [review]
